homeassistant/components/essent/sensor.py

```diff
@@ -0,0 +1,359 @@
+"""Sensor platform for Essent integration."""
+
+from __future__ import annotations
+
+from datetime import datetime
+from typing import Any
+
+from homeassistant.components.sensor import SensorDeviceClass, SensorEntity
+from homeassistant.const import CURRENCY_EURO
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+from homeassistant.util import dt as dt_util
+
+from .const import ENERGY_TYPE_ELECTRICITY, ENERGY_TYPE_GAS
+from .coordinator import EssentConfigEntry, EssentDataUpdateCoordinator
+from .entity import EssentEntity
+
+PARALLEL_UPDATES = 1
+
+
+def _parse_tariff_times(
+    tariff: dict[str, Any],
+) -> tuple[datetime | None, datetime | None]:
+    """Parse tariff start/end times and ensure they are timezone-aware."""
+    start_str = tariff.get("startDateTime")
+    end_str = tariff.get("endDateTime")
+
+    # Type narrowing for parse_datetime
+    start = dt_util.parse_datetime(start_str) if isinstance(start_str, str) else None
+    end = dt_util.parse_datetime(end_str) if isinstance(end_str, str) else None
+
+    if start and start.tzinfo is None:
+        start = dt_util.as_local(start)
+    if end and end.tzinfo is None:
+        end = dt_util.as_local(end)
+
+    return start, end
+
+
+def _format_dt_str(value: str | None) -> str | None:
+    """Format a datetime string as local ISO, falling back to original."""
+    if not value:
+        return None
+    parsed = dt_util.parse_datetime(value)
+    if not parsed:
+        return value
+    if parsed.tzinfo is None:
+        parsed = dt_util.as_local(parsed)
+    return parsed.isoformat()
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: EssentConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up Essent sensors."""
+    coordinator: EssentDataUpdateCoordinator = entry.runtime_data
+
+    entities = [
+        EssentCurrentPriceSensor(coordinator, ENERGY_TYPE_ELECTRICITY),
+        EssentNextPriceSensor(coordinator, ENERGY_TYPE_ELECTRICITY),
+        EssentAveragePriceSensor(coordinator, ENERGY_TYPE_ELECTRICITY),
+        EssentLowestPriceSensor(coordinator, ENERGY_TYPE_ELECTRICITY),
+        EssentHighestPriceSensor(coordinator, ENERGY_TYPE_ELECTRICITY),
+        EssentCurrentPriceSensor(coordinator, ENERGY_TYPE_GAS),
+        EssentNextPriceSensor(coordinator, ENERGY_TYPE_GAS),
```

Comment:
I'd recommend looking into entity descriptions. They're great for grouping entities with a similar behaviour with a generic entity, for which you change the appearance and the data it needs to source from. Check for appearances of `value_fn` in the code for an example

Reply:
Really cool way of defining sensors!